### ISSUE TITLE
[need improve] Branch miss in interpreter mode

### ISSUE THREAD

**Issue by eat4toast** [WasmEdge/WasmEdge#1457]
## Description
refer to code lib/executor/instantiate/function.cpp  https://github.com/WasmEdge/WasmEdge/blob/70ed44a14664f030e6c63ba20ec12e1ce16bc74b/lib/executor/instantiate/function.cpp#L25-L30 add atomic var to count branch, and found interpreter mode always run in **else** branch, AOT mode always run in **if** branch. Which seems to cause a huge branch miss with interpreter mode.(I am not sure the hardware things like branch prediction to record history will solve this issue)
### Current State
```
# ./WasmEdge/build/tools/wasmedge/wasmedge --dir .:. miniruby hello.rb
233
count if: 0 count else: 5781
# ./WasmEdge/build/tools/wasmedge/wasmedgec miniruby miniruby.wasm
[2022-05-16 23:41:31.263] [info] compile start
-----
# ./WasmEdge/build/tools/wasmedge/wasmedge --dir .:. miniruby.wasm hello.rb
233
count if: 5781 count else: 0
# cat hello.rb
puts(233)
#
```
### Expected

The newer version will automatically run with two style codes, which may reduce unnecessary branch misses.

## Environment

- Hardware Architecture: x64
- Operating system: Ubuntu 20.04

## Steps to Reproduce
```
// add follow header in file lib/executor/instantiate/function.cpp , and modify code same as below(insert atomic count)
#include <iostream>
#include <atomic>
...
Expect<void> Executor::instantiate(Runtime::Instance::ModuleInstance &ModInst,
                                   const AST::FunctionSection &FuncSec,
                                   const AST::CodeSection &CodeSec) {
std::atomic_int cnt_if{}, cnt_else{};
  // Get the function type indices.
  auto TypeIdxs = FuncSec.getContent();
  auto CodeSegs = CodeSec.getContent();

  // Iterate through the code segments to instantiate function instances.
  for (uint32_t I = 0; I < CodeSegs.size(); ++I) {
    // Create and add the function instance into the module instance.
    auto *FuncType = *ModInst.getFuncType(TypeIdxs[I]);
    if (auto Symbol = CodeSegs[I].getSymbol()) {
      ModInst.addFunc(&ModInst, *FuncType, std::move(Symbol));
      cnt_if++;
    } else {
      ModInst.addFunc(&ModInst, *FuncType, CodeSegs[I].getLocals(),
                      CodeSegs[I].getExpr().getInstrs());
      cnt_else++;
    }
  }
  std::cout<<"count if: "<<cnt_if<<" count else: "<<cnt_else<<'\n';
  return {};
}
...

```

**Actually run without miniruby just run with **hello.wasm** still got the same result.**

**Comment by q82419** [comment]
Hint:

In the header `include/loader/loader.h`, the input format (WASM, universal WASM, or shared library format) information can be found in these flags:

```
  bool IsSharedLibraryWASM;
  bool IsUniversalWASM;
```

Maybe the `WasmEdge::AST::Module` class can save the loaded format information for improve this when instantiation.

**Comment by hydai** [comment]
Hi @eat4toast 
Would you like to finish this little modification? Looks like we can check the first item and then bypass the remaining part.

**Comment by eat4toast** [comment]
I open a draft pr https://github.com/WasmEdge/WasmEdge/pull/1582  .

The detailed test results in this [whole project](https://github.com/eat4toast/fix_addfunc_test_result/blob/main/initialize_test/log).
In short, the cost time after modification seems greater than before, even though the branch count is really great.

I am not sure if this is related to my incorrect modification. 
(I doubt the loop and branch I write looks bad, which impact the compiler done right optimization ?)

**Comment by hydai** [comment]
Hi @eat4toast 
I would like to know if there is a comparison between the current master and your modification based on the current master?

**Comment by eat4toast** [comment]
Well, do you mean the detailed result comparison? In [this repo]( https://github.com/eat4toast/fix_addfunc_test_result/tree/main/initialize_test), the origin log stands for the current master and the new log stands for the after modification. You might compare from the named file if you have time.

> Hi @eat4toast I would like to know if there is a comparison between the current master and your modification based on the current master?

**Comment by eat4toast** [comment]
Hi, @hydai 
Thank you take time review. I test the way you recommend code snippet, and it works well! 
The time not enlarge again and even has slight shorter in my test case.

The test result: [0622_test](https://github.com/eat4toast/fix_addfunc_test_result/tree/main/0622_test)
The new pr: https://github.com/WasmEdge/WasmEdge/pull/1587

(SORRY, since my incorrect git usage, the origin draft pr is broken: 
I clone with --depth=1, and later using git commit --amend to modify commit , and then the other commit history will be rewrite)

**Comment by hydai** [comment]
Could you summarize the improvement in the description of the PR? So we can have better information about how much the PR improves.

**Comment by eat4toast** [comment]
> Could you summarize the improvement in the description of the PR? So we can have better information about how much the PR improves.

Sure, please take a look in https://github.com/WasmEdge/WasmEdge/pull/1587#issuecomment-1163895634
